### PR TITLE
Implement adaption prompt from Llama-Adapter paper

### PR DESCRIPTION
**Pull request by yeoedward**
Issue #235 
Paper: https://arxiv.org/abs/2303.16199
Official Implementation: https://github.com/ZrrSkywalker/LLaMA-Adapter

This implementation is quite specific to Llama although in principle the same technique can be used for other transformer models. It is implemented by replacing some of the `LlamaAttention` modules with `AdaptedAttention` modules that wrap the original ones, but bypass the original `forward()` so that we can inject trainable tokens and a gate. I can add a notebook example in a subsequent PR when I have the time.

### CODE REVIEW

**Comment by yeoedward** [comment]
Fixed `black` formatting and merge conflicts.

**Review by younesbelkada — COMMENTED** [review]
Thanks a mile for the PR! I think it's in a great shape! 
Can you run the styling checks ?
```bash
make style && make quality
```

**Comment by yeoedward** [comment]
CI test failures were caused by Llama dependencies not being available in earlier versions of the `transformers` library.
To ensure backwards compatibility, I've done the following:
1. Added import guards for llama classes from `transformers`. This will allow using `peft` to be used with earlier versions of `transformers` without breaking.
2. Skip unit tests that require those classes if the classes are not found. CI tests should pass.

**Comment by yeoedward** [comment]
Rebased on main and added support for multi-adapters

**Review by pacman100 — COMMENTED** [review]
This is awesome ✨! Great work @yeoedward 🔥. Thank you for blazing fast PR, adding multi-adapter support and adding tests. 

Overall comments:
1. The only drawback of the current implementation is being tightly coupled to the Llama model making it less generic. 
2. When I was thinking of the implementation of this method, I imagined `target_modules` pointing to the attention module, something like `self_attn` instead of `layers` (which we can get as parent module via `_get_submodules` method.
3. To make the implementation generic, having `attention_target_modules` as another config which is a `List[nn.Module]`.  The `AdaptedAttention` module's new forward should call the normal Attention layer's forward without any changes (so no modified_forward). Then use the `attention_target_modules` to get the `query_states`, `adapter_k` and `adapter_v`. Here, if there is only single fused_qkv as in gpt-2/bloom, just use `outputs.split(3, dim=2)` and do the necessary assignments.

**Comment by pacman100** [comment]
>  Recompute query_states, which will require importing the Llama-specific apply_rotary_pos_emb() (less duplication of code but with the perf penalty of recomputation).

I am leaning towards this as this would be generic and recomputation penalty would be minimal.
There can be a model to postprocessing function mapping which could be used for things like `apply_rotary_pos_emb `. Refer this code block for one such example:

https://github.com/huggingface/peft/blob/main/src/peft/peft_model.py#L246-L249

**Comment by yeoedward** [comment]
@pacman100 I've made the changes you suggested and rebased on main.

**Comment by winglian** [comment]
@yeoedward have you been able to validate the implementation results against the paper's results? I gave your branch a quick whirl with the same hyper parameters as the paper, but it definitely doesn't seem to converge as quickly.

**Comment by yeoedward** [comment]
@winglian I have validated the implementation on smaller problems that run on a CPU but haven't tried reproducing the paper's results at full problem size. When you say it doesn't converge as quickly, are you referring to number of epochs trained? Happy to look into it if you can provide a notebook / script, although I don't have access to 8 x A100s.

File: src/peft/tuners/adaption_prompt.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        if kwargs.get("output_attention", False):
+            raise NotImplementedError("output_attention is not currently supported.")
+
+        output, _, _ = self.model(**kwargs)
+        bsz = output.shape[0]
+        q_len = output.shape[1]
+
+        # (bsz, num_heads, adapter_len, head_dim)
+        adapter_k = (
+            self.model.k_proj(self.adaption_prompt)
```
Comment (pacman100) [inline src/peft/tuners/adaption_prompt.py:309]:
```suggestion
            key
```

File: src/peft/tuners/adaption_prompt.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+            .view(1, self.adapter_len, self.model.num_heads, self.model.head_dim)
+            .repeat(bsz, 1, 1, 1)
+            .transpose(1, 2)
+        )
+        # (bsz, num_heads, adapter_len, head_dim)
+        adapter_v = (
+            self.model.v_proj(self.adaption_prompt)
```
Comment (pacman100) [inline src/peft/tuners/adaption_prompt.py:316]:
```suggestion
            value
```

File: src/peft/tuners/adaption_prompt.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+        # (bsz, num_heads, q_len, adapter_len)
+        scores = torch.matmul(query_states, adapter_k.transpose(2, 3)) / math.sqrt(self.model.head_dim)
+        # Upcast attention to fp32
+        # (bsz, num_heads, q_len, adapter_len)
+        scores = self.adaption_gate * F.softmax(scores, dim=-1, dtype=torch.float32).to(query_states.dtype)
+        # (bsz, q_len, num_heads * head_dim)
+        adapter_output = torch.matmul(scores, adapter_v).transpose(1, 2).reshape(bsz, q_len, -1)
+        # (bsz, q_len, hidden_size)
+        adapter_output = self.model.o_proj(adapter_output)
```
Comment (pacman100) [inline src/peft/tuners/adaption_prompt.py:335]:
```suggestion
        if o_proj_layer is not None:
            adapter_output = getattr(self.model, o_proj_layer)(adapter_output)
```

File: src/peft/tuners/adaption_prompt.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+    position_ids = kwargs.get("position_ids")
+    bsz, q_len, _ = hidden_states.size()
+    query_states = model.q_proj(hidden_states).view(bsz, q_len, model.num_heads, model.head_dim).transpose(1, 2)
+    value_states = model.v_proj(hidden_states).view(bsz, q_len, model.num_heads, model.head_dim).transpose(1, 2)
+    cos, sin = model.rotary_emb(value_states, seq_len=q_len)
+    return llama_apply_rotary_pos_emb(query_states, cos, sin, position_ids)
+
+
+# Contains the config that is specific to a transformers model type.
+ModelTypeConfig = namedtuple("ModelTypeConfig", ["compute_query_states", "target_modules"])
```
Comment (pacman100) [inline src/peft/tuners/adaption_prompt.py:78]:
```suggestion
ModelTypeConfig = namedtuple("ModelTypeConfig", ["compute_query_states", "target_modules", "k_proj_layer", "v_proj_layer", "o_proj_layer"])
```

File: src/peft/tuners/adaption_prompt.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+# Mapping of transformers model types to their specific configuration.
+TRANSFORMERS_MODEL_CONFIG = {
+    "llama": ModelTypeConfig(compute_query_states=llama_compute_query_states, target_modules="self_attn"),
```
Comment (pacman100) [inline src/peft/tuners/adaption_prompt.py:81]:
```suggestion
    "llama": ModelTypeConfig(compute_query_states=llama_compute_query_states, target_modules="self_attn", k_proj_layer="k_proj", v_proj_layer="v_proj", o_proj_layer="o_proj"),
```

File: src/peft/tuners/adaption_prompt.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        """
+        Forward pass for the adapter which wraps the original LlamaAttention module.
+
+        "Official" paper implementation:
+        https://github.com/ZrrSkywalker/LLaMA-Adapter/blob/41c3546fe1997ab8a65809dc8d8f9252b19d9faf/llama/model.py#L141
+
+        Args:
+            kwargs: See the original LlamaAttention module.
+        """
+        if kwargs.get("use_cache", False):
```
[the remark this answers is not in the document]
Comment (pacman100) [reply src/peft/tuners/adaption_prompt.py:298]: Just passing remaining outputs as is should do work

**Review by pacman100 — COMMENTED** [review]
Hello @yeoedward, thank you for iterating. i have left comments

File: src/peft/tuners/adaption_prompt.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yeoedward) [reply src/peft/tuners/adaption_prompt.py]: My original intention was to avoid introducing additional complexity, but I think you're right it's less work than I thought. I've added support for it along with a simple unit test.

**Comment by yeoedward** [comment]
@pacman100 I've made the suggested changes and rebased on main.

**Review by pacman100 — APPROVED** [review]
Thank you @yeoedward  for iterating 🤗. This is awesome and you have done a great job writing extensive tests to cover many aspects of Adaption Prompt implementation. Thank you so much for putting in all the effort, LGTM! 🚀

**Comment by pacman100** [comment]
Could you please run `make style` and `make quality` to fix the quality issues?

**Comment by yeoedward** [comment]
@pacman100 I couldn't reproduce the style error (which was in a file unrelated to this PR), but I rebased on `main` and fixed some conflicts, which perhaps was the reason for the style checker failing. I've re-run the unit tests and quality checks, so hopefully this is good to go. Thanks for reviewing!

**Comment by pacman100** [comment]
Thank you @yeoedward! 🤗

**Comment by tathagata-raha** [comment]
It took me a bit of time to find that the paper has been implemented, given that it is present in the list of the Supported methods. However, I haven't seen any documentation in the Tuners page [here](https://huggingface.co/docs/peft/package_reference/tuners)
